microprofile/cdi/src/main/java/io/helidon/microprofile/cdi/ContainerInstanceHolder.java

```diff
@@ -40,35 +44,52 @@ static void set(HelidonContainer container) {
     }
 
     // return true if the container was reset, indicating somebody started CDI by hand and then shut it down
-    static synchronized boolean isReset() {
-        return isReset;
+    static boolean isReset() {
+        return access(() -> isReset);
     }
 
-    static synchronized HelidonContainer get() {
-        HelidonContainer helidonContainer = CONTAINER.get();
-        if (null == helidonContainer) {
-            helidonContainer = fromBuildTimeInitializer();
-            CONTAINER.compareAndSet(null, helidonContainer);
-        }
-        return helidonContainer;
+    static HelidonContainer get() {
+        return access(() -> {
+            HelidonContainer helidonContainer = CONTAINER.get();
+            if (null == helidonContainer) {
+                helidonContainer = fromBuildTimeInitializer();
+                CONTAINER.compareAndSet(null, helidonContainer);
```

Comment:
Naïve reader surprise moment: normally you check to see if a `compareAndSet` "worked", and if it did not, then you know that the value changed via another thread.  I was expecting, again, just rolling along as a reader:
```
HelidonContainer helidonContainer = CONTAINER.get();
if (helidonContainer == null) {
    helidonContainer = fromBuildTimeInitializer();
    if (!CONTAINER.compareAndSet(null, helidonContainer)) {
        return CONTAINER.get();
    }
}
return helidonContainer;
```
Additionally, if `CONTAINER` is an `AtomicReference` (too lazy to muck around in the Github interface to verify this) then there shouldn't be any locking/synchronization needed beyond it, right? On the other hand, if there are normally other things that go on when it is set, then there's no need for it to be an `AtomicReference`.  I recognize this is not code that you've added; just strange semantics IMHO.

Reply:
You raise good points. I consciously have tried to avoid changing other code during this work...letting sleeping dogs lie and all that.